rfc/20250211-s3-locking-with-conditional-writes.md

```diff
@@ -0,0 +1,148 @@
+* HTTP 412 (Precondition Failed) - Means that the locking file is already there. Therefore, the whole process should exit because the lock couldn't be acquired.
+* HTTP 409 (Conflict) - Means that the there was a conflict of concurrent requests. AWS recommends to retry the request in such cases, but we could just handle this similarly to the 412 case.
+
+#### Digest updates
+The S3 conditional writes, offer also the option to overwrite an object conditionally, only when the `ETag` of the existing object is matching the `ETag` that it is given in the `If-Match: <etag from reading>` header.
+
+We can use this option to write the digest (`<state-file-path>-md5`) object after the state file is saved.
+
+To do so, when [reading the digest](https://github.com/opentofu/opentofu/blob/6614782/internal/backend/remote-state/s3/client.go#L88) for checking against the state file, we should store also the `ETag` of the read file.
+Later, when writing the newly generated digest, we can use the `ETag` in the `PutObjectInput.IfMatch` to ensure that the file didn't get any other update in the meantime:
+* If the action succeeds, we should just proceed.
+* If the action fails (404, 409, 412), we could warn the user and retry the writing of the digest object without the `ETag`.
+  * The retry should be done to ensure consistency between the state file content and its digest.
+
+As a short pseudocode, this would look something like:
+```go
+state := getState()
+digest, etag := getMD5()
+// ... process
+err := putMD5(sum, etag)
+if err {
+	warn("concurrent writes on the state digest detected")
+	err := putMD5(sum, "") // Empty ETag should rewrite the digest object
+}
+```
+
+### Open Questions
+
+* Do we want to provide the option to have the lock objects into another bucket? This will break the feature parity.
+* When writing the digest (*-md5) object, should we consider conditional writing by using the `IfMatch` argument?
+  ```go
+  func (actor S3Actions) UpdateDigest(ctx context.Context, bucket string, key string, beforeApplyDigest string) error {
+      newDigest := uuid.NewString()
+      input := &s3.PutObjectInput{
+          Bucket:            aws.String(bucket),
+          Key:               aws.String(key),
+          Body:              bytes.NewReader([]byte(newDigest)),
+          ChecksumAlgorithm: types.ChecksumAlgorithmSha256,
+          IfMatch:           aws.String(beforeApplyDigest),
+      }
+      _, err := actor.S3Client.PutObject(ctx, input)
+      return err
+  }
+  ```
+  Asking because this could also lead to other issues in case it doesn't match. I would suggest to have this only for showing an error in case the digest is different.
+* Right now, since the MD5 digest is stored in DynamoDB, this is written only when the `dynamodb_table` is specified.
+  * Should we do the same in the new implementation? Write the MD5 digest only if the `use_lockfile=true`?
```

Comment:
I think we should only write lock files if the locking is enabled, so I would say yes.

Reply:
Removed some questions since are not needed anymore. More details in the [above comment](https://github.com/opentofu/opentofu/pull/2511#discussion_r1955828819).